### PR TITLE
Lstm learning to main

### LINKED ISSUE #270: Add LSTM policy
To enhance the RL method currently implemented in ASSUME, we want to add diffrent policies and buffers. 

This includes an LSTM policy for the actor and/or critic instead of the plain NN to have some time awareness. 

Furthermore, a new buffer is implemented, which uses not random sampling but importance sampling instead.

### PR DESCRIPTION
**Pull request by kim-mskw**
# Pull Request

## Related Issue
Closes #270

## Description
Enable the LSTM policies instead of the fully connected NN for RL in the MATD3 algorithm 

## Testing
Created a new environment and ran the examples and notebooks

## Checklist
Please check all applicable items:

- [x] Code changes are sufficiently documented (docstrings, inline comments, `doc` folder updates)
- [x] New unit tests added for new features or bug fixes
- [x] Existing tests pass with the changes
- [x] Reinforcement learning examples are operational (for DRL-related changes)
- [x] Code tested with both local and Docker databases
- [x] Code follows project style guidelines and best practices
- [ ] Changes are backwards compatible, or deprecation notices added
- [x] New dependencies added to `pyproject.toml`
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included
- [x] Consent to release this PR's code under the GNU Affero General Public License v3.0

## Additional Notes (if applicable)
[Any additional information, concerns, or areas you want reviewers to focus on]

## Screenshots (if applicable)
[Add screenshots to demonstrate visual changes]

### CODE REVIEW

File: assume/reinforcement_learning/network_architecture.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    def forward(self, obs):
+        x = F.relu(self.FC1(obs))
+        x = F.relu(self.FC2(x))
+        x = F.softsign(self.FC3(x))
+        # x = th.tanh(self.FC3(x))
+
+        return x
+
+
+class LSTMActor(nn.Module):
```
Comment (nick-harder) [inline assume/reinforcement_learning/network_architecture.py:124]: nice work! We need to couple this somehow with the learning strategies so the values are forwarded for the number of time series and single observations. Having it like this hard coded is not the best thing, and I believe if we leave it like this now, we won't fix it for a long time.

File: assume/reinforcement_learning/network_architecture.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+    def __init__(self, obs_dim: int, act_dim: int, float_type):
+        super().__init__()
+        self.float_type = float_type
+        self.no_of_timeseries = 2  # TODO: variable no. of input timeseries
+        self.forecast_horizon = int(
+            (obs_dim - 2) / self.no_of_timeseries
+        )  # TODO: variable number of additional input features, currently 2 (marginal costs, capacity)
+
+        self.LSTM1 = nn.LSTMCell(self.no_of_timeseries, 8, dtype=float_type)
```
Comment (nick-harder) [inline assume/reinforcement_learning/network_architecture.py:140]: not for now, but maybe some food for thoughts, we could just try this out with a GRU units instead of LSTM, which for our use case would probably be sufficient and much faster.

File: assume/reinforcement_learning/network_architecture.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        for time_step in x1.split(1, dim=2):
+            time_step = time_step.reshape(-1, 2)
+            h_t, c_t = self.LSTM1(time_step, (h_t, c_t))
+            h_t2, c_t2 = self.LSTM2(h_t, (h_t2, c_t2))
+            outputs += [h_t2]
+
+        outputs = th.cat(outputs, dim=1)
+        x = th.cat((outputs, x2), dim=1)
+
+        x = F.relu(self.FC1(x))
```
Comment (nick-harder) [inline assume/reinforcement_learning/network_architecture.py:181]: did you try to use here after the FC1 a dropot of like 0.2 maybe? just wondering, not for this PR. We could test it out with the mlp and this one if this would improve the generalization

File: assume/scenario/loader_csv.py [hunk 4]
```diff
@@ -760,6 +760,7 @@ def run_learning(
         "avg_all_eval": [],
         "episodes_done": 0,
         "eval_episodes_done": 0,
+        "noise_scale": world.learning_config.get("noise_scale", 1.0),
```
Comment (nick-harder) [inline assume/scenario/loader_csv.py:763]: oh, I didn't even realize we didn't have this in the config...thanks!

File: assume/strategies/learning_strategies.py [hunk 5]
```diff
@@ -46,7 +49,17 @@ class RLStrategy(LearningStrategy):
     """
 
     def __init__(self, *args, **kwargs):
-        super().__init__(obs_dim=50, act_dim=2, unique_obs_dim=2, *args, **kwargs)
+        obs_dim = kwargs.get("observation_dimension", 50)
```
Comment (nick-harder) [inline assume/strategies/learning_strategies.py:52]: the idea here was that these dimensions are fixed for this specific strategy, as they are fixed by the create_obs and create_actions functions. Defining these values from config won't work since the code has to be actively adjusted to accomodate this

**Review by nick-harder — CHANGES_REQUESTED** [review]
thanks for the nice work, looks good! I have left some comments to fix before we can merge

File: assume/common/units_operator.py [hunk 6]
```diff
@@ -488,3 +490,201 @@ async def formulate_bids(
...
+                receiver_id=learning_role_id,
+                receiver_addr=learning_role_addr,
+                content={
+                    "context": "rl_training",
+                    "type": "replay_buffer",
+                    "data": rl_agent_data,
+                },
+            )
+
+    def write_learning_params(
```
[the remark this answers is not in the document]
Comment (kim-mskw) [reply assume/common/units_operator.py:651]: of course! I did not expect the merging to behave in that way

File: assume/common/utils.py [hunk 7]
```diff
@@ -471,3 +472,27 @@ def convert_to_rrule_freq(string: str) -> tuple[int, int]:
     freq = freq_map[string[-1]]
     interval = int(string[:-1])
     return freq, interval
+
+
+def rename_study_case(path: str, old_key: str, new_key: str):
```
[the remark this answers is not in the document]
Comment (kim-mskw) [reply assume/common/utils.py:477]: This was used for several runs of the same study, which enabled an analasis of the stochastitciity of the learned behavior. I think it is therefore useful.

File: examples/inputs/example_02a/config.yaml [hunk 8]
```diff
@@ -14,6 +14,7 @@ tiny:
     trained_policies_save_path: null
     max_bid_price: 100
     algorithm: matd3
+    network_architecture: mlp
```
[the remark this answers is not in the document]
Comment (kim-mskw) [reply examples/inputs/example_02a/config.yaml:17]: fair changed it

File: assume/reinforcement_learning/algorithms/__init__.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
+# SPDX-FileCopyrightText: ASSUME Developers
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+from torch import nn
+
+from assume.reinforcement_learning.network_architecture import MLPActor, LSTMActor
+
+policy_aliases: dict[str, type[nn.Module]] = {
```
[the remark this answers is not in the document]
Comment (kim-mskw) [reply assume/reinforcement_learning/algorithms/__init__.py:9]: fair, chaged it

File: assume/reinforcement_learning/learning_role.py [hunk 10]
```diff
@@ -77,7 +78,7 @@ def __init__(
             learning_config.get("episodes_collecting_initial_experience", 5), 1
         )
 
-        self.train_freq = learning_config.get("train_freq", 1)
+        self.train_freq = learning_config.get("train_freq", "1h")
```
[the remark this answers is not in the document]
Comment (kim-mskw) [reply assume/reinforcement_learning/learning_role.py:81]: It is still used for the gradient steps?

File: compose.yml [hunk 11]
```diff
@@ -19,7 +19,7 @@ services:
       # needed for normal image
       #- ./assume-db:/var/lib/postgresql/data
       # /home/postgres/data is path for timescaledev image
-      - ./assume-db:/home/postgres/pgdata
+      - ./assume-db:/home/postgres/data
```
[the remark this answers is not in the document]
Comment (kim-mskw) [reply compose.yml:22]: It was the only way it worked on the extrernal machine, but I'll change it back

[the remark this answers is not in the document]
Comment (maurerle) [reply compose.yml]: I think you sometimes need to create the folder `mkdir assume-db` yourself, with user permission before changing this, because the timescaledb container behaves weirdly, I think..

Changing the path to something unused or commenting it out would lead to data being lost if the container is recreated - but other than that works of course as well

File: assume/reinforcement_learning/network_architecture.py (hunk 3 above)
Comment (mthede) [reply assume/reinforcement_learning/network_architecture.py]: that's currently in our backlog, we haven't tried it yet. but we will keep it in mind!

File: assume/reinforcement_learning/network_architecture.py (hunk 2 above)
Comment (mthede) [reply assume/reinforcement_learning/network_architecture.py]: yes, we thought about GRU (or xLSTM) as well. right now we're not focussing on implementing another variation of the actor network.

File: assume/reinforcement_learning/algorithms/__init__.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mthede) [reply assume/reinforcement_learning/algorithms/__init__.py]: renamed the "policy_class" as well into "neural_network_architecture_class", so that its clear that it's referring to the same thing (but it's a class object instead of the string/alias). I'm not too happy with the long name though. Would you prefer policy_class although the naming wouldn't be fully consistent? 

Another possibility would be: "actor_architecture" (e.g. "mlp"/"lstm"), "actor_class" (-> MLPActor/LSTMActor) and "actor_aliases" to match it. I think it would be also clearer that the config option only changes the neural network architecture of the actor, but not of the critics.

File: assume/strategies/learning_advanced_orders.py [hunk 12]
```diff
@@ -582,11 +596,13 @@ def load_actor_params(self, load_path):
 
         params = th.load(directory, map_location=self.device)
 
-        self.actor = Actor(self.obs_dim, self.act_dim, self.float_type)
+        self.actor = self.policy_class(self.obs_dim, self.act_dim, self.float_type)
         self.actor.load_state_dict(params["actor"])
```
Comment (mthede) [inline assume/strategies/learning_advanced_orders.py:600]: @kim-mskw please check if that's necessary. I think "unique_obs_dim" and "num_timeseries_obs_dim" would be needed for LSTM with advanced orders. Therefore copied the actor instantiation from TD3's create_actors() function. ".to(self.device)" might not be needed?

File: assume/strategies/learning_strategies.py [hunk 13]
```diff
@@ -470,11 +484,13 @@ def load_actor_params(self, load_path):
 
         params = th.load(directory, map_location=self.device)
 
-        self.actor = Actor(self.obs_dim, self.act_dim, self.float_type)
+        self.actor = self.policy_class(self.obs_dim, self.act_dim, self.float_type)
         self.actor.load_state_dict(params["actor"])
```
Comment (mthede) [inline assume/strategies/learning_strategies.py:488]: @kim-mskw same question here as for load_actor_params() in advanced orders

File: assume/reinforcement_learning/network_architecture.py (hunk 1 above)
Comment (mthede) [reply assume/reinforcement_learning/network_architecture.py]: fixed this, unique observations and number of timeseries are now forwarded from learning strategy.

File: assume/strategies/learning_advanced_orders.py (hunk 12 above)
Comment (kim-mskw) [reply assume/strategies/learning_advanced_orders.py:606]: yes it makes sense to have these here now!
The .to(self.deivce) is used to move a model or tensor to a specified device, which can be either a CPU or a GPU. Why would it not be needed here?

File: assume/reinforcement_learning/learning_role.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kim-mskw) [reply assume/reinforcement_learning/learning_role.py:83]: And the gradient steps are used in the loop of the maTD3. So I guess we still need it.

File: assume/strategies/learning_advanced_orders.py (hunk 12 above)
Comment (mthede) [reply assume/strategies/learning_advanced_orders.py:606]: '.to(self.device)' wasn't used before and I'm not sure if line 597 'params = th.load(directory, map_location=self.device)' therefore already covers this part.

Comment (kim-mskw) [reply assume/strategies/learning_advanced_orders.py:606]: Ah understood, from my perspective it is however necesary that the actor model itself is on the device and not only the params.

Comment (kim-mskw) [reply assume/strategies/learning_advanced_orders.py:606]: We also use it for the critics.

**Comment by kim-mskw** [comment]
# Pull Request

## Related Issue
Closes #[issue number] (if applicable)

## Description


## Changes Proposed


## Testing
Created a new environment and ran the examples and notebooks

## Checklist
Please check all applicable items:

- [ ] Code changes are sufficiently documented (docstrings, inline comments, `doc` folder updates)
- [ ] New unit tests added for new features or bug fixes
- [ ] Existing tests pass with the changes
- [ ] Reinforcement learning examples are operational (for DRL-related changes)
- [ ] Code tested with both local and Docker databases
- [ ] Code follows project style guidelines and best practices
- [ ] Changes are backwards compatible, or deprecation notices added
- [ ] New dependencies added to `pyproject.toml`
- [ ] A note for the release notes `doc/release_notes.rst` of the upcoming release is included
- [ ] Consent to release this PR's code under the GNU Affero General Public License v3.0

## Additional Notes (if applicable)
[Any additional information, concerns, or areas you want reviewers to focus on]

## Screenshots (if applicable)
[Add screenshots to demonstrate visual changes]

File: assume/reinforcement_learning/algorithms/__init__.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mthede) [reply assume/reinforcement_learning/algorithms/__init__.py]: yes, the architecture itself could be used. however, it would be necessary to implement another critic (child) class with the same architecture.

[the remark this answers is not in the document]
Comment (mthede) [reply assume/reinforcement_learning/algorithms/__init__.py]: Kim and I agreed to change the naming to actor_architecture - analog critic_architecture would be possible, but is currently not implemented

**Review by nick-harder — APPROVED** [review]
